lib/logstash/string_interpolation.rb

```diff
@@ -0,0 +1,155 @@
+require "thread_safe"
+require "forwardable"
+
+module LogStash
+  class StringInterpolation
+    # Floats outside of these upper and lower bounds are forcibly converted
+    # to scientific notation by Float#to_s
+    MIN_FLOAT_BEFORE_SCI_NOT = 0.0001
+    MAX_FLOAT_BEFORE_SCI_NOT = 1000000000000000.0
+
+    INSTANCE = StringInterpolation.new
+    CACHE = ThreadSafe::Cache.new
+    TEMPLATE_TAG_REGEXP = /%\{[^}]+\}/
+
+    def evaluate(event, template)
+      if template.is_a?(Float) and
+          (template < MIN_FLOAT_BEFORE_SCI_NOT or template >= MAX_FLOAT_BEFORE_SCI_NOT) then
+        return template = ("%.15f" % template).sub(/0*$/,"")
+      end
+      
+      template = template.to_s
+
+      if not_cachable?(template)
+        return template
+      else
+        compiled = CACHE.get_or_default(template, nil) || CACHE.put(template, compile_template(template))
+        compiled.evaluate(event)
+      end
+    end
+
+    def self.evaluate(event, format)
+      INSTANCE.evaluate(event, format)
+    end
+
+    private
+    def not_cachable?(template)
+      template.index("%").nil?
```

Comment:
this is food for thoughts: since this is static, we could annotate these template at config "compile time" and avoid doing this costly string index check at every interpolation